### ISSUE TITLE
The v0.7 release

### ISSUE THREAD

**Issue by anyzelman** [Algebraic-Programming/ALP#65]
This issue keeps track of new features that will make it into v0.7:
 - a vertex-centric Pregel API (#64),
 - automatic HyperDAG extraction.

The following will hopefully make it into v0.7:
 - a port of the nonblocking backend by Mastoras et al. [1] on top of the latest develop.

[1] https://ieeexplore.ieee.org/document/9835271

Feel welcome to suggest new features or bug fixes for a v0.7 release in this thread.

**Comment by djelovina** [comment]
@anyzelman 
From my side I can propose GMRES and AMG  merge into develop. Both branches need a new rebase..

**Comment by anyzelman** [comment]
> @anyzelman From my side I can propose GMRES and AMG merge into develop. Both branches need a new rebase..

GMRES OK, please ping me one of these days on which branch where should be rebased and then marged. For AMG, can you check sync with Alberto what's happing on that side of things? We currently have, if nothing changes, AMG, HPCG, and CG, all in disjoint implementations while a lot of functionality is shared (and also causing at least in part some of the confusing readings for the nonblocking backend on AMG).

**Comment by djelovina** [comment]
Indeed, there is some code duplication (also as you mentioned with the CG algorithm), which has to be cleaned. I discussed this with Alberto, the plan is to merge AMG branch into develop once his new version of HPCG is already merged. GMRES can go independently, I could also write an input matrix  generator, rather than reading from the input file.

**Comment by anyzelman** [comment]
That is OK for me but it likely means missing the v0.7 blocker. Is that OK for you? (You’re racing against Aristeidis;)

发件人: djelovina ***@***.***
发送时间: 29 November 2022 16:09
收件人: Algebraic-Programming/ALP ***@***.***>
抄送: Albert Jan YZELMAN ***@***.***>; Assign ***@***.***>
主题: Re: [Algebraic-Programming/ALP] The v0.7 release (Issue #65)


Indeed, there is some code duplication (also as you mentioned with the CG algorithm), which has to be cleaned. I discussed this with Alberto, the plan is to merge AMG branch into develop once his new version of HPCG is already merged. GMRES can go independently, I could also write an input matrix generator, rather than reading from the input file.

**Comment by anyzelman** [comment]
> Sorry, I did not get what do you mean by the last message :)

I think we discussed it offline, but just to keep this thread up to date: I meant once Aristeidis' non-blocking backend is ready, we will enter a free for releasing v0.7:)

**Comment by anyzelman** [comment]
Remaining blockers before starting release engineering of v0.7: #156 #169

**Comment by anyzelman** [comment]
Remaining blocker: #169

**Comment by anyzelman** [comment]
All blockers resolved -- starting the below loop:

Associated branch(es) to the current release candidates should be named `vx.y.z-rcn`, where vx.y.z corresponds to the version ID in the title and `n` is the release candidate ID.

Looped start (loop count is `n`):
- [ ] Bump new version number in ./CMakeLists.txt
- [ ] Set same version in doxygen (doxy.conf)
- [ ] Check compiler warnings and suppressions
- [ ] Remove trailing spaces and tabs, and check for spaces-before-tabs
- [ ] Update changelog.md according to MRs in develop after rc`n-1` and the current rc`n`. (Let rc`-1` be `develop`.)
- [ ] Test rc`n` (see below for ``sub-checklist'')

If unsuccessful:
- fix bugs in separate issues / MRs that merge into develop
- rebase vx.y.z-rc`n` unto develop and rename it to vx.y.z-rc`n+1`
- uncheck all checked boxes above and restart loop

If successful:
- [ ] merge rc`n` into develop
- [ ] ensure internal CI remains happy
- [ ] merge develop into master
- [ ] ensure internal CI remains happy
- [ ] tag master after merge with version ID, and push the new tag

Testing TODOs:
- [ ] GitHub CI should flag OK
- [ ] Internal CI should flag OK (warning: the branch name matters!)
- [ ] Ubuntu ARM without LPF without banshee
- [ ] Ubuntu ARM with LPF and MPICH without banshee
- [ ] Ubuntu x86 with LPF and MPICH without banshee
- [ ] Fedora x86 without LPF without banshee
- [ ] Fedora x86 with LPF and MPICH without banshee
- [ ] Any OS on any arch with LPF and OpenMPI without banshee
- [ ] CentOS x86 with LPF and MPICH without banshee

**Comment by anyzelman** [comment]
Fedora on x86 with Linux kernel 5.8.18, GCC 9.3.1, and MPICH 3.3.2: smoke and unit tests OK

**Comment by anyzelman** [comment]
Testing on ARM found a minor issue, fixed in rc2. Restarting loop

**Comment by anyzelman** [comment]
- Fedora on x86 with Linux kernel 5.8.18, GCC 9.3.1, and MPICH 3.3.2: smoke and unit tests OK
- Ubuntu on x86 with Linux kernel 5.4.0, GCC 9.4.0, and MPICH 3.3.2: smoke and unit tests OK
- Ubuntu on ARM with Linux kernel 5.4.0, GCC 9.4.0, and MPICH 3.3.2: smoke and unit tests OK

**Comment by anyzelman** [comment]
Up to now not much variety with the GCC versions: splicing in one additional mandatory testing check using the latest GCC.

**Comment by anyzelman** [comment]
Fedora on x86 with Linux kernel 5.8.18, GCC 9.3.1, and only reference backends enabled: smoke and unit tests OK

**Comment by anyzelman** [comment]
Ubuntu on ARM with Linux kernel 5.4.0, GCC 9.4.0, without LPF: smoke and unit tests OK

**Comment by anyzelman** [comment]
Fedora on x86 with Linux kernel 5.8.18, GCC 12.2.0, without LPF: smoke and unit tests OK

**Comment by anyzelman** [comment]
Changelog does not require an update-- the bug that was fixed in RC3 is again the reference(_omp) eWiseApply, and again a rarely (or never) trigged bug, which is already summarised in the changelog. The support for GCC 12, OTOH, will be mentioned in the release notes. Will keep RC3's commit history intact and push them unto `develop`.

**Comment by anyzelman** [comment]
v0.7 was released a while ago, therefore this one can close=)
